### PR TITLE
[CURATOR-311] SharedValue could hold stall data when quourm membership changes

### PR DESCRIPTION
**Pull request by oza**
Adding error handling to refresh stale data when stateChanged is called.

### CODE REVIEW

File: curator-recipes/src/main/java/org/apache/curator/framework/recipes/shared/SharedValue.java [hunk 1]
```diff
@@ -75,14 +75,41 @@ public void process(WatchedEvent event) throws Exception
...
+
+            case RECONNECTED:
+                // update currentValue
+                try {
+                    state.compareAndSet(State.STALE, State.STARTING);
+                    startInternal();
+                    state.compareAndSet(State.STARTING, State.STARTED);
+                } catch (Exception e) {
+                    ThreadUtils.checkInterrupted(e);
+                    log.error("Could not re-start instances after reconnection", e);
```
Comment (oza) [inline curator-recipes/src/main/java/org/apache/curator/framework/recipes/shared/SharedValue.java]: I would like to know whether this error handling is correct. Please let me know if you know a better way.

**Comment by Randgalt** [comment]
I'm -1 on this PR overall. The purpose of `state` internally is not to hold the connection state. It's to hold the started state of the instanced. Mixing the two is bound to cause problems. 

A simpler solution would be:

```java
    private final ConnectionStateListener connectionStateListener = new ConnectionStateListener()
    {
        @Override
        public void stateChanged(CuratorFramework client, ConnectionState newState)
        {
            notifyListenerOfStateChanged(newState);
            if ( newState == ConnectionState.RECONNECTED )
            {
                try
                {
                    readValueAndNotifyListenersInBackground();
                }
                catch ( Exception e )
                {
                    log.error("Could not read value after reconnect", e);
                }
            }
        }
    };
```

File: curator-recipes/src/test/java/org/apache/curator/framework/recipes/shared/TestSharedCount.java [hunk 2]
```diff
@@ -389,4 +390,93 @@ public void processResult(CuratorFramework client, CuratorEvent event) throws Ex
             CloseableUtils.closeQuietly(curatorFramework);
         }
     }
+
+    @Test
+    public void testDisconnectReconnectEventDoesNotFireValueWatcherWithMultipleClients() throws Exception
```
Comment (Randgalt) [inline curator-recipes/src/test/java/org/apache/curator/framework/recipes/shared/TestSharedCount.java]: I don't think this tests the problem. I tried this test with an unchanged version SharedValue and it passes.

**Comment by oza** [comment]
Thanks Jordan for taking a look. I will double check my test case and resend fixed version later.

**Comment by oza** [comment]
Based on the report in jira, I will create a test case, TestSharedValueWithCluster.java, with TestingCluster to reproduce the problem. Please let me know if you know a better way.

**Comment by oza** [comment]
@Randgalt I could create a test case to reproduce the problem by injecting emulated faulty watcher instead of using pseudo cluster. Could you check it?

File: curator-recipes/src/test/java/org/apache/curator/framework/recipes/shared/TestSharedCount.java [hunk 3]
```diff
@@ -381,12 +385,118 @@ public void processResult(CuratorFramework client, CuratorEvent event) throws Ex
             }).forPath("/count");
             flushDone.await(5, TimeUnit.SECONDS);
 
-            Assert.assertEquals(2, numChangeEvents.get());
+            Assert.assertEquals(numChangeEvents.get(), 3);
```
Comment (oza) [inline curator-recipes/src/test/java/org/apache/curator/framework/recipes/shared/TestSharedCount.java]: This means this PR includes semantic change of countHasChanged callback: after reconnecting, countHasChanged() is called.

Comment (Randgalt) [inline curator-recipes/src/test/java/org/apache/curator/framework/recipes/shared/TestSharedCount.java]: This commit is no longer available. Can you post it again?

Comment (oza) [inline curator-recipes/src/test/java/org/apache/curator/framework/recipes/shared/TestSharedCount.java]: Sure.

**Comment by oza** [comment]
@Randgalt can you see changes?

**Comment by Randgalt** [comment]
No - the commit is still missing.

**Comment by oza** [comment]
Hmm, I don't know what is happening. I would like to close this and try to create new PR.

**Pull request closed without merging**